scrape/scrape.go

```diff
@@ -419,22 +454,47 @@ func (sp *scrapePool) sync(targets []*Target) {
 			sp.activeTargets[hash] = t
 			sp.loops[hash] = l
 
-			go l.run(interval, timeout, nil)
+			uniqueTargets[hash] = l
 		} else {
+			uniqueTargets[hash] = nil
 			// Need to keep the most updated labels information
 			// for displaying it in the Service Discovery web page.
 			sp.activeTargets[hash].SetDiscoveredLabels(t.DiscoveredLabels())
 		}
 	}
 
+	targetScrapePoolTargetsPostRelabeling.WithLabelValues(sp.config.JobName).Set(float64(len(uniqueTargets)))
+	if sp.config.TargetsLimit > 0 && len(uniqueTargets) > int(sp.config.TargetsLimit) {
+		sp.droppedByLimitTargets = make([]*Target, len(uniqueTargets))
+		i := 0
+		for hash, l := range uniqueTargets {
+			t := sp.activeTargets[hash]
+			sp.droppedByLimitTargets[i] = t
+			if l != nil {
+				l.cancel()
```

Comment:
We should only ever cancel in-progress scrapes on shutdown, in all other situations we should wait for the scrape to complete.